experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts

```diff
@@ -88,12 +105,18 @@ export function getProxyForNode<TSchema extends TreeSchema>(
 	if (schemaIsLeaf(schema)) {
 		return treeNode.value as ProxyNode<TSchema>;
 	}
-	if (schemaIsFieldNode(schema)) {
-		return createListProxy(treeNode) as ProxyNode<TSchema>;
-	}
-	if (schemaIsStruct(schema)) {
-		return createObjectProxy(treeNode, schema) as ProxyNode<TSchema>;
+	const isFieldNode = schemaIsFieldNode(schema);
+	if (isFieldNode || schemaIsStruct(schema)) {
+		const cachedProxy = getCachedProxy(treeNode);
+		if (cachedProxy !== undefined) {
+			return cachedProxy as ProxyNode<TSchema>;
+		}
+
+		const proxy = isFieldNode ? createListProxy(treeNode) : createObjectProxy(treeNode, schema);
+		cacheProxy(treeNode, proxy);
+		return proxy as ProxyNode<TSchema>;
```

Comment:
If you combine function passing, type params, and '...args: unknown[]' in just the right way, you could probably rework getCachedProxy into:
```ts
if (schemaIsFieldNode(..)) {
    return getProxy(createListProxy, treeNode)
} else if (schemaIsStruct(..)) {
   return getProxy(createObjectProxy, treeNode, schema)
}
```
...but it's fine to commit as-is and I *might* come by afterword and obsessively fiddling with it to avoid checking isFieldNode twice.  :-)

(PS - I used to avoid passing functions out of a desire to avoid overhead from indirect calls, but after benchmarking some of Jack William's code, had to concede that V8 is surprisingly smart about inlining these.)

Reply:
That's a good idea to try. I'm all for obsessive fiddling, I just want to get everything working first :D Please fiddle away.